### PR TITLE
[ELI-20]: Pagination, file splitting, ui/ux updates, multiselect links, test updates.

### PR DESCRIPTION
**Pull request by aliakseiTraihel**
- spliting report file to files size by 500 items.
- pagination 
- added actual information about modified links
- ui/ux updates for fix broken and replace by pattern buttons
- ability to select multiple links for fix
- added run button for start job
- updates for tests

image
image

### CODE REVIEW

File: core/src/main/java/com/exadel/etoolbox/linkinspector/core/models/ui/PaginationModel.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+import org.apache.sling.models.annotations.DefaultInjectionStrategy;
+import org.apache.sling.models.annotations.Model;
+import org.apache.sling.models.annotations.injectorspecific.Self;
+import org.apache.sling.models.annotations.injectorspecific.SlingObject;
+
+import javax.annotation.PostConstruct;
+import java.util.Optional;
+
+/**
+ * Represents the model with pagination logic. Based on the number of pages in the current report.
```
Comment (smiakchilo) [inline core/src/main/java/com/exadel/etoolbox/linkinspector/core/models/ui/PaginationModel.java:18]: (wording) Did not get. who or what is based on the number?)

File: core/src/main/java/com/exadel/etoolbox/linkinspector/core/models/ui/PaginationModel.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+    public int getSize() {
+        return size;
+    }
+
+    public int getPreviousPage() {
+        return page - 1;
+    }
+
+    public int getPageBeforeOne() {
```
Comment (smiakchilo) [inline core/src/main/java/com/exadel/etoolbox/linkinspector/core/models/ui/PaginationModel.java:69]: (wording) Probably it is `getPageBeforePrevious`. "Before one" does not sound allright). Same is to "after one"

File: core/src/main/java/com/exadel/etoolbox/linkinspector/core/services/GridDataSource.java [hunk 3]
```diff
@@ -21,7 +21,10 @@ public interface GridDataSource {
      * Generates {@link DataSource} necessary for displaying items on the Link Inspector's page
      * within the grid (granite/ui/components/coral/foundation/table)
      *
+     * @param page - page number of report from request params
```
Comment (smiakchilo) [inline core/src/main/java/com/exadel/etoolbox/linkinspector/core/services/GridDataSource.java:24]: Just "current page number")

File: core/src/main/java/com/exadel/etoolbox/linkinspector/core/services/data/DataFeedService.java [hunk 4]
```diff
@@ -18,35 +18,57 @@
...
+import java.util.Map;
 
 /**
  * Provides the service responsible for managing the data feed based on the set of resources generated
  * by ${@link GridResourcesGenerator}
  */
 public interface DataFeedService {
+
+    /**
+     * The sling resource type of grid row items
```
Comment (smiakchilo) [inline core/src/main/java/com/exadel/etoolbox/linkinspector/core/services/data/DataFeedService.java:30]: (cosmetic) "Sling" is a brand name, therefore, must be capitalized.

File: core/src/main/java/com/exadel/etoolbox/linkinspector/core/services/data/impl/DataFeedServiceImpl.java [hunk 5]
```diff
@@ -142,59 +106,50 @@ public List<Resource> dataFeedToResources() {
...
-                } catch (JSONException e) {
-                    LOG.error("Failed to convert json object to GridResource", e);
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void modifyDataFeed(Map<String, String> propertyLocationLinkMap, int page) {
+        try (ResourceResolver serviceResourceResolver = repositoryHelper.getServiceResourceResolver()) {
+            List<GridResource> updatedResources = readGridResources(serviceResourceResolver, page).stream().peek(resource -> {
+                if (propertyLocationLinkMap.containsKey(resource.getPropertyLocation())) {
```
Comment (smiakchilo) [inline core/src/main/java/com/exadel/etoolbox/linkinspector/core/services/data/impl/DataFeedServiceImpl.java:126]: ❗ Please extract this code to a separate method. A lambda inside a lambda is..., ouch

File: core/src/main/java/com/exadel/etoolbox/linkinspector/core/services/helpers/CsvHelper.java [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+import com.exadel.etoolbox.linkinspector.core.models.ui.GridViewItem;
+import com.exadel.etoolbox.linkinspector.core.services.data.models.GridResource;
+import org.apache.sling.api.resource.ResourceResolver;
+
+import java.util.List;
+
+public interface CsvHelper {
+
+    /**
+     * Reading items from csv and adapt them to the {@link GridResource}
```
Comment (smiakchilo) [inline core/src/main/java/com/exadel/etoolbox/linkinspector/core/services/helpers/CsvHelper.java:12]: (grammar) "Reads... and adapts...". 
(spelling) (Here and below in the whole PR). CSV is an abbreviation and must be spelled in capital letters. Same is "JCR",

File: core/src/main/java/com/exadel/etoolbox/linkinspector/core/services/helpers/CsvHelper.java [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+     *
+     * @param resourceResolver - {@link ResourceResolver}
+     * @param page             - page number of csv report
+     * @return the {@link List<GridResource>} representing the List of extracted grid resources form csv
+     */
+    List<GridResource> readCsvReport(ResourceResolver resourceResolver, int page);
+
+    /**
+     * Creating csv reports by list of {@link GridViewItem}. Reports are splitting on parts and used as pages.
```
Comment (smiakchilo) [inline core/src/main/java/com/exadel/etoolbox/linkinspector/core/services/helpers/CsvHelper.java:21]: Same as above + "Reports are splittED IN parts". Please consider using a spellchecker for the whole text. "Grammarly" maybe... or some ChatGPT or something of that kind

File: core/src/main/java/com/exadel/etoolbox/linkinspector/core/services/helpers/CsvHelper.java [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+import java.util.List;
+
+public interface CsvHelper {
+
+    /**
+     * Reading items from csv and adapt them to the {@link GridResource}
+     *
+     * @param resourceResolver - {@link ResourceResolver}
+     * @param page             - page number of csv report
+     * @return the {@link List<GridResource>} representing the List of extracted grid resources form csv
```
Comment (smiakchilo) [inline core/src/main/java/com/exadel/etoolbox/linkinspector/core/services/helpers/CsvHelper.java:16]: (typo) "... **from** CSV"

File: core/src/main/java/com/exadel/etoolbox/linkinspector/core/services/util/LinkInspectorResourceUtil.java [hunk 9]
```diff
@@ -59,6 +52,36 @@ public static void removeResource(String path, ResourceResolver resourceResolver
...
+        try {
+            Session session = resourceResolver.adaptTo(Session.class);
+            if (session == null) {
+                LOG.warn("Session is null, recreating node is interrupted");
+                return;
+            }
+            JcrUtil.createPath(path, JcrConstants.NT_UNSTRUCTURED, session);
+            session.save();
+        } catch (RepositoryException e) {
+            LOG.error(String.format("Failed to recreate node %s", path), e);
```
Comment (smiakchilo) [inline core/src/main/java/com/exadel/etoolbox/linkinspector/core/services/util/LinkInspectorResourceUtil.java:65]: Until Java 17, String formatter operations are quite inefficient. Please avoid them whenever you can. In the current case, Message Formatter works better:
```
LOG.error("Failed to recreate node {}", path, e);
```

File: core/src/main/java/com/exadel/etoolbox/linkinspector/core/services/util/LinkInspectorResourceUtil.java [hunk 10]
```diff
@@ -59,6 +52,36 @@ public static void removeResource(String path, ResourceResolver resourceResolver
...
+        }
+    }
+
+    public static void addParamToNode(String path, ResourceResolver resourceResolver, String paramName, Long paramValue) {
```
Comment (smiakchilo) [inline core/src/main/java/com/exadel/etoolbox/linkinspector/core/services/util/LinkInspectorResourceUtil.java:69]: Why `paramValue` is a boxed Long, suddenly?)) Use `long`. You can use `int` allright, this is no problem, and then no need to convert to long in [CsvHelperImpl.java](https://github.com/exadel-inc/etoolbox-link-inspector/pull/28/files?diff=unified&w=0#diff-4de93ea8a9319a21106040b10e20a94df2692dfd3082ae236d9d6687be396a85)

File: core/src/main/java/com/exadel/etoolbox/linkinspector/core/servlets/DownloadReportServlet.java [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+                    CsvUtil.convertPageNumberToPath(DataFeedService.CSV_REPORT_NODE_PATH, i)
+            );
+            if (report != null) {
+                records.addAll(CsvUtil.readCsvItems(report.adaptTo(InputStream.class), CsvUtil.CSV_COLUMNS));
+            }
+        }
+        return records;
+    }
+
+    private String[] getValues(CSVRecord record) {
```
Comment (smiakchilo) [inline core/src/main/java/com/exadel/etoolbox/linkinspector/core/servlets/DownloadReportServlet.java:81]: (cosmetic) You can make this method return `Object[]` just from the start , and then no need to convert to Object[] in line 58

**Review by smiakchilo — APPROVED** [review]

File: core/src/main/java/com/exadel/etoolbox/linkinspector/core/services/util/LinkInspectorResourceUtil.java (hunk 10 above)
Comment (aliakseiTraihel) [reply core/src/main/java/com/exadel/etoolbox/linkinspector/core/services/util/LinkInspectorResourceUtil.java:69]: Origianly we don't have int type in jcr

File: core/src/main/java/com/exadel/etoolbox/linkinspector/core/services/helpers/CsvHelper.java (hunk 7 above)
Comment (aliakseiTraihel) [reply core/src/main/java/com/exadel/etoolbox/linkinspector/core/services/helpers/CsvHelper.java:21]: Fine. A good tip is to use ChatGPT.

File: core/src/main/java/com/exadel/etoolbox/linkinspector/core/services/util/ServletUtil.java [hunk 12]
```diff
@@ -37,10 +42,23 @@ public static String getRequestParamString(SlingHttpServletRequest request, Stri
...
+                .collect(Collectors.toList()))
+                .orElse(Collections.emptyList());
+    }
+
     public static boolean getRequestParamBoolean(SlingHttpServletRequest request, String param) {
         return Boolean.parseBoolean(getRequestParamString(request, param));
     }
 
+    public static int getRequestParamInt(SlingHttpServletRequest request, String param) {
+        String path = getRequestParamString(request, param);
```
Comment (vihnatsenka) [inline core/src/main/java/com/exadel/etoolbox/linkinspector/core/services/util/ServletUtil.java:58]: "path" variable name is incorrect.

**Pull request merged**